diff --git a/parallel_hashmap/phmap.h b/parallel_hashmap/phmap.h
index 37f0f7c..3b1fe51 100644
--- a/parallel_hashmap/phmap.h
+++ b/parallel_hashmap/phmap.h
@@ -1212,10 +1212,11 @@ public:
             std::swap(growth_left(), that.growth_left());
             std::swap(infoz_, that.infoz_);
         } else {
-            reserve(that.size());
+            throw std::runtime_error("FIXME");
+            // reserve(that.size());
             // Note: this will copy elements of dense_set and unordered_set instead of
             // moving them. This can be fixed if it ever becomes an issue.
-            for (auto& elem : that) insert(std::move(elem));
+            // for (auto& elem : that) insert(std::move(elem));
         }
     }
 
